Import libraries.

In [1]:
import pandas as pd    # data formatting
import numpy as np     # numeric library
from sklearn.neighbors import KNeighborsClassifier  # machine learning
from sklearn.externals import joblib # exporting model
from sklearn.metrics import confusion_matrix
import random

In [2]:
%matplotlib inline
from matplotlib import pylab, pyplot  # plotting

Read data.

In [42]:
data = pd.read_csv('seed_data/scores.csv')

Pivot data.

In [43]:
d = data.pivot(index='producer_id', columns='performer_id', values='score')

In [44]:
d.head()

performer_id  1        2        3        4        5        13       18       \
producer_id                                                                   
70                  0        0        0        0        0        0        0   
72                  1        1        0        1        1        1        0   
82                  0        1        1        0        0        0        0   
110                 0        1        1        1        1        1        0   
112                 0        1        1        1        0        0        0   

performer_id  20       21       22       ...  1730289  1731488  1731529  \
producer_id                              ...                              
70                  1        0        0  ...        0        0        0   
72                  1        0        0  ...        0        0        0   
82                  1        0        1  ...        0        0        0   
110                 0        0        0  ...        0        0        0   
112                 0        0        0  ...        0        0        0   

performer_id  1732680  1732985  1734410  1734423  1734890  1735951  1736560  
producer_id                                                                  
70                  0        0        0        0        0        0        0  
72                  0        0        0        0        0        0        0  
82                  0        0        0        0        0        0        0  
110                 0        0        0        0        0        0        0  
112                 0        0        0        0        0        0        0  

[5 rows x 9665 columns]

This dataframe has 307 rows and 9665 columns.  Each row corresponds to 1 producer.  Each column corresponds to a performer.

In [45]:
d.shape

(307, 9665)

In [46]:
d.tail()

performer_id  1        2        3        4        5        13       18       \
producer_id                                                                   
1646412             0        0        0        0        0        0        0   
1663511             0        0        0        0        0        0        0   
1687278             0        0        0        0        0        0        0   
1702507             0        0        0        0        0        0        0   
1706060             0        0        0        0        0        0        0   

performer_id  20       21       22       ...  1730289  1731488  1731529  \
producer_id                              ...                              
1646412             0        0        0  ...        0        0        0   
1663511             0        0        0  ...        0        0        0   
1687278             0        0        0  ...        0        0        0   
1702507             0        0        0  ...        0        0        0   
1706060             0        0        0  ...        0        0        0   

performer_id  1732680  1732985  1734410  1734423  1734890  1735951  1736560  
producer_id                                                                  
1646412             0        0        0        0        0        0        0  
1663511             0        0        0        0        0        0        0  
1687278             0        0        0        0        0        0        0  
1702507             0        0        0        0        0        0        0  
1706060             0        0        0        0        0        0        0  

[5 rows x 9665 columns]

In [47]:
d.describe()

performer_id     1           2           3           4           5        \
count         307.000000  307.000000  307.000000  307.000000  307.000000   
mean            0.039088    0.127036    0.130293    0.228013    0.026059   
std             0.194121    0.333557    0.337175    0.420236    0.159570   
min             0.000000    0.000000    0.000000    0.000000    0.000000   
25%             0.000000    0.000000    0.000000    0.000000    0.000000   
50%             0.000000    0.000000    0.000000    0.000000    0.000000   
75%             0.000000    0.000000    0.000000    0.000000    0.000000   
max             1.000000    1.000000    1.000000    1.000000    1.000000   

performer_id     13          18          20          21          22       ...  \
count         307.000000  307.000000  307.000000  307.000000  307.000000  ...   
mean            0.159609    0.003257    0.058632    0.013029    0.055375  ...   
std             0.366841    0.057073    0.235318    0.113585    0.229083  ...   
min             0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%             0.000000    0.000000    0.000000    0.000000    0.000000  ...   
50%             0.000000    0.000000    0.000000    0.000000    0.000000  ...   
75%             0.000000    0.000000    0.000000    0.000000    0.000000  ...   
max             1.000000    1.000000    1.000000    1.000000    1.000000  ...   

performer_id     1730289     1731488     1731529     1732680     1732985  \
count         307.000000  307.000000  307.000000  307.000000  307.000000   
mean            0.003257    0.003257    0.003257    0.003257    0.003257   
std             0.057073    0.057073    0.057073    0.057073    0.057073   
min             0.000000    0.000000    0.000000    0.000000    0.000000   
25%             0.000000    0.000000    0.000000    0.000000    0.000000   
50%             0.000000    0.000000    0.000000    0.000000    0.000000   
75%             0.000000    0.000000    0.000000    0.000000    0.000000   
max             1.000000    1.000000    1.000000    1.000000    1.000000   

performer_id     1734410     1734423     1734890     1735951     1736560  
count         307.000000  307.000000  307.000000  307.000000  307.000000  
mean            0.003257    0.003257    0.003257    0.003257    0.003257  
std             0.057073    0.057073    0.057073    0.057073    0.057073  
min             0.000000    0.000000    0.000000    0.000000    0.000000  
25%             0.000000    0.000000    0.000000    0.000000    0.000000  
50%             0.000000    0.000000    0.000000    0.000000    0.000000  
75%             0.000000    0.000000    0.000000    0.000000    0.000000  
max             1.000000    1.000000    1.000000    1.000000    1.000000  

[8 rows x 9665 columns]

Show performer's 5 nearest neighbors.

In [64]:
knn = KNeighborsClassifier(n_neighbors=5)

In [65]:
list(d.index)

[70,
 72,
 82,
 110,
 112,
 123,
 146,
 158,
 170,
 184,
 217,
 224,
 337,
 383,
 491,
 509,
 577,
 662,
 685,
 686,
 698,
 699,
 848,
 862,
 1056,
 1070,
 1529,
 1553,
 1555,
 1954,
 2024,
 2194,
 2260,
 2268,
 2524,
 2726,
 2906,
 3264,
 3608,
 3646,
 4463,
 4617,
 4709,
 4736,
 5214,
 5898,
 6021,
 6708,
 7122,
 7161,
 7462,
 7561,
 7570,
 7573,
 7856,
 7898,
 9476,
 9662,
 10233,
 10418,
 11546,
 11845,
 12908,
 12991,
 13046,
 13585,
 13872,
 13945,
 14166,
 14766,
 14777,
 15708,
 15870,
 16678,
 17102,
 17598,
 17746,
 18277,
 18379,
 18595,
 19824,
 19829,
 20185,
 21188,
 21456,
 22499,
 22698,
 23132,
 23315,
 23691,
 24099,
 24434,
 24741,
 26234,
 27258,
 27259,
 27438,
 27469,
 27472,
 27473,
 27474,
 27486,
 27500,
 27511,
 27606,
 27611,
 27612,
 27626,
 27635,
 27639,
 27640,
 27663,
 27665,
 27670,
 27689,
 27765,
 27794,
 27801,
 27819,
 27822,
 27869,
 27938,
 27987,
 28305,
 28455,
 28628,
 28938,
 29780,
 29935,
 29947,
 29978,
 30206,
 30761,
 31168,
 31695,
 3194

In [66]:
knn.fit(d.iloc[:,:], d.index)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [67]:
predictions = knn.predict(d.iloc[:,:])

In [68]:
joblib.dump(knn, "trained-model_producers.pkl")

['trained-model_producers.pkl']

List index 0 = performer, following list indices are that performer's nearest neighbors.

In [69]:
dist, ind = knn.kneighbors(d.loc[112,:].values.reshape(1, -1))
print(ind[0])

[  4  51 238 254  91]


In [70]:
dist

array([[ 0.        , 14.56021978, 14.76482306, 14.76482306, 14.79864859]])

In [71]:
[list(d.index)[i] for i in ind[0]]

[112, 7561, 649283, 984361, 24434]